File: grazi/src/main/java/tanvd.grazi/GraziApplicationSettings.java
package tanvd.grazi;

import com.intellij.openapi.components.*;
import org.jetbrains.annotations.*;
import tanvd.grazi.grammar.*;
import tanvd.grazi.language.*;
import tanvd.grazi.spellcheck.*;

import java.io.*;
import java.util.*;

@State(name = "GraziApplicationSettings", storages = @Storage("grazi.xml"))
public class GraziApplicationSettings implements PersistentStateComponent<GraziApplicationSettings.State> {
    private State myState = new State();

    public static GraziApplicationSettings getInstance() {
        return ServiceManager.getService(GraziApplicationSettings.class);
    }

    @NotNull
    @Override
    public State getState() {
        return myState;
    }

    @Override
    public void loadState(@NotNull State state) {
        myState = state;

        init();
    }

    void init() {
        LangChecker.INSTANCE.clear();
        SpellDictionary.Companion.setGraziFolder(myState.graziFolder);
        GrammarEngine.Companion.getInstance().setEnabledLangs(new ArrayList<>(myState.languages));
    }

    public static class State {
        final Set<String> languages = new HashSet<>();
        public File graziFolder = new File(System.getProperty("user.home"), ".grazi");

        {
            languages.add("en");
        }
    }
}

File: grazi/src/main/kotlin/tanvd/grazi/ui/GraziApplicationSettings.kt
package tanvd.grazi.ui

import com.intellij.openapi.components.*
import tanvd.grazi.grammar.GrammarEngineService
import java.util.*

@State(name = "GraziApplicationSettings", storages = [Storage("grazi.xml")])
class GraziApplicationSettings : PersistentStateComponent<GraziApplicationSettings.State> {
    data class State(val languages: HashSet<String> = hashSetOf("en"))

    private var _state = State()

    val myState: State
        get() = _state

    override fun getState(): State {
        return _state
    }

    override fun loadState(state: State) {
        _state = state
    }

    fun loadLanguages() {
        GrammarEngineService.getInstance().enabledLangs = ArrayList(_state.languages)
    }

    companion object {
        val instance: GraziApplicationSettings
            get() = ServiceManager.getService(GraziApplicationSettings::class.java)
    }
}
